bin/problem.py

```diff
@@ -343,56 +343,58 @@ def build_program(p):
+            check_constraints: True if the validao
+
+        Returns:
+            False: something went wrong
+            OutputValidator if validator_type is 'output'
+            list(Validator) otherwise
+        """
+        assert validator_type in ['input', 'answer', 'output']
 
         key = (validator_type, check_constraints)
         if key in problem._validators:
             return problem._validators[key]
+        ok = True
 
-        # For default 'output' validation, use default_output_validator.cpp.
-        if validator_type == 'output' and problem.settings.validation == 'default':
-            validators = [
-                validate.OutputValidator(
-                    problem, config.tools_root / 'support' / 'default_output_validator.cpp'
-                )
-            ]
-            bar = ProgressBar(f'Build {validator_type} validators', items=validators)
-            ok = True
-            for p in validators:
-                bar.start(p)
-                ok &= p.build(bar)
-                bar.done()
-            bar.finalize(print_done=False)
-            if not ok:
-                validators = False
-            problem._validators[key] = validators
-            return validators
-
-        validator_dir = 'input' if validator_type == 'input_format' else 'output'
-
-        paths = glob(problem.path / (validator_dir + '_validators'), '*') + glob(
-            problem.path / (validator_dir + '_format_validators'), '*'
-        )
-
-        if len(paths) == 0:
-            # Only log/warn missing validators in generate mode.
-            if config.args.action == 'generate':
-                if validator_type == 'output_format':
-                    log(f'No {validator_type} validators found.')
-                else:
-                    warn(f'No {validator_type} validators found.')
-                    problem._validators[key] = False
-                    return False
-        if validator_type == 'output' and problem.interactive and len(paths) != 1:
-            error(
-                f'Found {len(paths)} output validators, but validation type {problem.settings.validation} needs exactly one.'
-            )
-            problem._validators[key] = False
-            return False
+        subdirs_for_type = {
+                'answer': ['answer_validators', 'answer_format_validators'],
+                'output': ['output_validator', 'output_validators'],
+                'input': ['input_validators', 'input_format_validators'],
+                }
+        paths_for_type = {
+                vtype: [g for sdir in sdirs for g in glob(problem.path / sdir, '*')]
+                for vtype, sdirs in subdirs_for_type.items()
+                }
+
+        # Handle default output validation
+        if problem.settings.validation == 'default':
+            if paths_for_type['output']:
+                error("Validation is default but custom output validator exists")
+                ok = False
```

Comment:
Hmm; so in this case we do build the default output validator but then anyway end up returning `False`. Can we just bail early? Or do you think it's ugly to have to repeatedly update the cache to `False` here and elsewhere we have `ok = False`?
(Or maybe the code following here already doesn't really do much anyway once `ok=False` so it's fine.)

Reply:
We _could_ bail early, but one of the first things I did to the code (in particular because I wanted to make sure the caching worked) is to only return at the end. It’s prettier and safer this way.